source/opt/loop_descriptor.h

```diff
@@ -0,0 +1,246 @@
+  inline const_iterator cbegin() const { return nested_loops_.begin(); }
+  inline const_iterator cend() const { return nested_loops_.end(); }
+
+  // Returns the header (first basic block of the loop). This block contains the
+  // OpLoopMerge instruction.
+  inline BasicBlock* GetHeaderBlock() { return loop_header_; }
+  inline const BasicBlock* GetHeaderBlock() const { return loop_header_; }
+
+  // Returns the latch basic block (basic block that holds the back-edge).
+  inline BasicBlock* GetLatchBlock() { return loop_continue_; }
+  inline const BasicBlock* GetLatchBlock() const { return loop_continue_; }
+
+  // Returns the BasicBlock which marks the end of the loop.
+  inline BasicBlock* GetMergeBlock() { return loop_merge_; }
+  inline const BasicBlock* GetMergeBlock() const { return loop_merge_; }
+
+  // Returns the loop pre-header, nullptr means that the loop predecessor does
+  // not qualify as a preheader.
+  // The preheader is the unique predecessor that:
+  //   - Dominates the loop header;
+  //   - Has only the loop header as successor.
+  inline BasicBlock* GetPreHeaderBlock() { return loop_preheader_; }
+
+  // Returns the loop pre-header.
+  inline const BasicBlock* GetPreHeaderBlock() const { return loop_preheader_; }
+
+  // Returns true if this loop contains any nested loops.
+  inline bool HasNestedLoops() const { return nested_loops_.size() != 0; }
+
+  // Returns the depth of this loop in the loop nest.
+  // The outer-most loop has a depth of 1.
+  inline size_t GetDepth() const {
+    size_t lvl = 1;
+    for (const Loop* loop = GetParent(); loop; loop = loop->GetParent()) lvl++;
+    return lvl;
+  }
+
+  // Adds |nested| as a nested loop of this loop. Automatically register |this|
+  // as the parent of |nested|.
+  inline void AddNestedLoop(Loop* nested) {
+    assert(!nested->GetParent() && "The loop has another parent.");
+    nested_loops_.push_back(nested);
+    nested->SetParent(this);
+  }
+
+  inline Loop* GetParent() { return parent_; }
+  inline const Loop* GetParent() const { return parent_; }
+
+  inline bool HasParent() const { return parent_; }
+
+  // Returns true if this loop is itself nested within another loop.
+  inline bool IsNested() const { return parent_ != nullptr; }
+
+  // Returns the set of all basic blocks contained within the loop. Will be all
+  // BasicBlocks dominated by the header which are not also dominated by the
+  // loop merge block.
+  inline const BasicBlockListTy& GetBlocks() const {
+    return loop_basic_blocks_;
+  }
+
+  // Returns true if the basic block |bb| is inside this loop.
+  inline bool IsInsideLoop(const BasicBlock* bb) const {
+    return IsInsideLoop(bb->id());
+  }
+
+  // Returns true if the basic block id |bb_id| is inside this loop.
+  inline bool IsInsideLoop(uint32_t bb_id) const {
+    return loop_basic_blocks_.count(bb_id);
+  }
+
+  // Adds the Basic Block |bb| this loop and its parents.
+  void AddBasicBlockToLoop(const BasicBlock* bb) {
```

Comment:
Seems like this should check that |bb| is actually dominated by the header and postdominated by the merge blocks (in the case of structured loops)

Reply:
I added this method as public so a client could update the loop descriptor (to preserve the analysis). In such situation, the dominator tree might not be updated but that can be a requirement.

The check should be an assert right ?